### PR TITLE
Bicaws7 360/asn editable field

### PR DESCRIPTION
**Pull request by tausifpatel**
### [Exception Handler - ASN (Editable Field)](https://dsdmoj.atlassian.net/jira/software/c/projects/BICAWS7/boards/1368?selectedIssue=BICAWS7-360)

Make ASN field editable for the following scenarios
- The case must have at least one exception
- Current user has the permission to edit exceptions
- The case is locked to the current user
- Error status is unresolved.
- Case is in Phase 1 (Hearing outcome)

### CODE REVIEW

File: src/features/CourtCaseDetails/Tabs/Panels/DefendantDetails.tsx [hunk 1]
```diff
@@ -13,38 +13,98 @@ import ErrorPromptMessage from "components/ErrorPromptMessage"
...
+  const [isValidAsn, setIsValidAsn] = useState<boolean>(true)
+  const handleAsnChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
+    const asn = e.target.value.toUpperCase()
+    setIsValidAsn(isAsnFormatValid(asn))
+    amendFn("asn")(asn)
+  }
+
+  const asnFormGroupError = isValidAsn ? "" : "govuk-form-group--error"
+  const isAsnEditable =
+    courtCase.canUserEditExceptions && courtCase.aho.Exceptions.length > 0 && courtCase.phase === Phase.HEARING_OUTCOME
```
Comment (C-gyorfi) [inline src/features/CourtCaseDetails/Tabs/Panels/DefendantDetails.tsx:61]: `courtCase.aho.Exceptions.length > 0`

 don't we need to check if it is an ASN exception?

Comment (tausifpatel) [reply src/features/CourtCaseDetails/Tabs/Panels/DefendantDetails.tsx:61]: No. As long as there is an unresolved exception, we want to make it editable.

File: src/services/dto/courtCaseDto.ts [hunk 2]
```diff
@@ -13,6 +18,10 @@ export const courtCaseToDisplayPartialCourtCaseDto = (courtCase: CourtCase): Dis
     errorLockedByUsername: courtCase.errorLockedByUsername,
     errorReport: courtCase.errorReport,
     errorStatus: courtCase.errorStatus ?? null,
+    canUserEditExceptions:
```
Comment (C-gyorfi) [inline src/services/dto/courtCaseDto.ts:21]: I wonder if `canUserEditExceptions` could be defined elsewhere, it feels strange to pass current users as a dependency for court case DTO

File: src/features/CourtCaseDetails/Tabs/Panels/DefendantDetails.tsx (hunk 1 above)
Comment (C-gyorfi) [reply src/features/CourtCaseDetails/Tabs/Panels/DefendantDetails.tsx:61]: Then I think its redundant, `.canUserEditExceptions` already checks if error status is unresolved(should only be unresolved if there are unresolved exceptions) 
```
  const isAsnEditable =
    courtCase.canUserEditExceptions && courtCase.phase === Phase.HEARING_OUTCOME
```

File: src/services/dto/courtCaseDto.ts (hunk 2 above)
Comment (richrace) [reply src/services/dto/courtCaseDto.ts:21]: You've got everything you need before you get to the UI (still on the server). So it's encapsulated the DTO and you don't need to expensive lookups.

Comment (richrace) [reply src/services/dto/courtCaseDto.ts:21]: So I think that's the right place for it be.

File: src/features/CourtCaseDetails/Tabs/Panels/DefendantDetails.tsx (hunk 1 above)
Comment (tausifpatel) [reply src/features/CourtCaseDetails/Tabs/Panels/DefendantDetails.tsx]: `courtCase.aho.Exceptions.length > 0`  is removed and updated an e2e test

**Review by C-gyorfi — DISMISSED** [review]

**Review by richrace — APPROVED** [review]

**Pull request merged**